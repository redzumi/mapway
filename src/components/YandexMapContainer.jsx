import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styled from 'styled-components'

class YandexMapContainer extends Component {
  componentDidMount = async () => {
    await this.createMap()
    this.addMapEvents()
    this.createPlacemarksCollection()
    this.createPolyline()
  }

  componentWillReceiveProps(nextProps) {
    nextProps.placemarks.forEach((item, index) => {
      if (this.props.placemarks[index] !== item) this.updatePlacemarkInCollection(item, index)
    })

    if (this.props.placemarks.length <= nextProps.placemarks.length) return

    this.placemarksCollection.splice(
      nextProps.placemarks.length,
      this.props.placemarks.length - nextProps.placemarks.length
    )
    this.placemarksStringGeometry.splice(
      nextProps.placemarks.length,
      this.props.placemarks.length - nextProps.placemarks.length
    )

    this.updatePlacemarksIcons()
  }

  getAddressByCoordinates = async coordinates => {
    const objectsByCoordinates = await window.ymaps.geocode(coordinates, {
      results: 1,
    })
    return objectsByCoordinates.geoObjects.get(0)
  }

  updatePlacemarksIcons = () => {
    const collectionLength = this.placemarksCollection.getLength()
    this.placemarksCollection.each((placemark, index) => {
      let placemarkIcon = 'islands#grayIcon'
      let placemarkIconContent = ''

      if (index === 0) {
        placemarkIcon = 'islands#redStretchyIcon'
        placemarkIconContent = 'Начало'
      }
      if (index === collectionLength - 1) {
        placemarkIcon = 'islands#darkBlueStretchyIcon'
        placemarkIconContent = 'Конец'
      }

      placemark.options.set('preset', placemarkIcon)
      placemark.properties.set('iconContent', placemarkIconContent)
    })
  }

  updatePlacemarkInCollection(data, index) {
    const placemark = new window.ymaps.GeoObject(
      {
        geometry: {
          type: 'Point',
          coordinates: data.coordinates,
        },
        properties: {
          balloonContent: data.name,
          hintContent: 'Загрузка...',
        },
      },
      {
        draggable: true,
        preset: 'islands#grayIcon',
      }
    )

    if (index > this.placemarksCollection.getLength() - 1) this.placemarksCollection.add(placemark)
    else this.placemarksCollection.set(index, placemark)

    this.placemarksStringGeometry.set(index, placemark.geometry.getCoordinates())

    this.getAddressByCoordinates(placemark.geometry.getCoordinates()).then(addressObject => {
      placemark.properties.set('hintContent', addressObject.getAddressLine())
    })

    placemark.events.add('mouseenter', () => {
      placemark.hint.open()
    })

    placemark.events.add('dragend', () => {
      this.props.onPlacemarkCoordinatesUpdate(index, placemark.geometry.getCoordinates())
    })

    placemark.events.add('drag', () => {
      this.placemarksStringGeometry.set(index, placemark.geometry.getCoordinates())
    })
  }

  addMapEvents() {
    this.map.events.add('actiontick', e => {
      const tick = e.get('tick')
      const center = this.map.options
        .get('projection')
        .fromGlobalPixels(tick.globalPixelCenter, tick.zoom)
      this.props.onMapCenterUpdate(center)
    })
  }

  createPolyline = () => {
    this.placemarksStringGeometry = new window.ymaps.geometry.LineString()

    const placemarksPolyline = new window.ymaps.Polyline(
      this.placemarksStringGeometry,
      {},
      {
        draggable: false,
        strokeColor: '#000000',
        strokeWidth: 4,
        strokeOpacity: 0.6,
      }
    )

    this.map.geoObjects.add(placemarksPolyline)
  }

  createPlacemarksCollection() {
    this.placemarksCollection = new window.ymaps.GeoObjectCollection(
      {},
      {
        preset: 'islands#redCircleIcon',
        strokeWidth: 4,
        geodesic: true,
      }
    )

    this.placemarksCollection.events.add('add', this.updatePlacemarksIcons)
    this.placemarksCollection.events.add('set', this.updatePlacemarksIcons)

    this.map.geoObjects.add(this.placemarksCollection)
  }

  createMap() {
    return new Promise(resolve => {
      window.ymaps.ready(() => {
        this.map = new window.ymaps.Map(
          this.mapElement,
          {
            center: this.props.mapCenter,
            zoom: 12,
            controls: [],
          },
          {
            buttonMaxWidth: 300,
          }
        )
        resolve()
      })
    })
  }

  render() {
    return (
      <Map
        innerRef={element => {
          this.mapElement = element
        }}
      />
    )
  }
}

YandexMapContainer.propTypes = {
  placemarks: PropTypes.arrayOf(
    PropTypes.shape({
      id: PropTypes.string.isRequired,
      name: PropTypes.string.isRequired,
      coordinates: PropTypes.array.isRequired,
    })
  ).isRequired,
  mapCenter: PropTypes.arrayOf(PropTypes.number, PropTypes.number).isRequired,
  onMapCenterUpdate: PropTypes.func.isRequired,
  onPlacemarkCoordinatesUpdate: PropTypes.func.isRequired,
}

const Map = styled.div`
  width: 600px;
  height: 400px;
  padding: 10px;
`

export { YandexMapContainer as default, Map }
