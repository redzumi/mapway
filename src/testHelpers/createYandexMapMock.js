const createYandexMapMock = onMapInitializationFinished => {
  class EventsCollection {
    add = () => {}
  }

  class GeoObjectCollection {
    constructor() {
      this.events = new EventsCollection()
    }

    add = () => {}
    set = () => {}
    getLength = () => {}
    each = () => {}
    splice = () => {}
  }

  class OptionsCollection {
    get = () => this
    fromGlobalPixels = () => 'test'
  }

  class LineString {
    set = () => {}
    splice = () => {}
  }

  class Geometry {
    LineString = () => new LineString()
    getCoordinates = () => {}
  }

  class Polyline {}

  class GeoObject {
    constructor() {
      this.geometry = new Geometry()
      this.events = new EventsCollection()
      this.properties = {
        get: () => {},
        set: () => {},
      }
      this.hint = {
        open: () => {},
      }
    }
  }

  class Map {
    constructor() {
      this.geoObjects = new GeoObjectCollection()
      this.events = new EventsCollection()
      this.options = new OptionsCollection()
    }
  }

  const geocode = async () => ({
    geoObjects: {
      get: () => ({
        getAddressLine: () => 'Test Address',
      }),
    },
  })

  return {
    ready: callback => {
      callback()
      onMapInitializationFinished()
    },
    geometry: new Geometry(),
    geocode,
    Map,
    GeoObject,
    GeoObjectCollection,
    Polyline,
  }
}

export default createYandexMapMock
