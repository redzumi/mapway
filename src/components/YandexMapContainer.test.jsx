import React from 'react'
import { mount } from 'enzyme'
import createYandexMapMock from '../testHelpers/createYandexMapMock'
import YandexMapContainer from './YandexMapContainer'

let mapComponent
beforeEach(done => {
  window.ymaps = createYandexMapMock(done)
  mapComponent = mount(
    <YandexMapContainer
      placemarks={[]}
      mapCenter={[55.75120138894101, 37.62371955158138]}
      onMapCenterUpdate={jest.fn()}
      onPlacemarkCoordinatesUpdate={jest.fn()}
    />
  )
})

afterEach(() => {
  mapComponent.unmount()
})

describe('YandexMapContainer', () => {
  it('renders correctly', () => {
    expect(mapComponent).toMatchSnapshot()
  })

  it('updates placemark after addition', () => {
    const componentInstance = mapComponent.instance()
    const spyUpdatePlacemark = jest.spyOn(componentInstance, 'updatePlacemarkInCollection')

    componentInstance.componentWillReceiveProps({ placemarks: [{ name: 'Test placemark 1' }] })

    expect(spyUpdatePlacemark).toBeCalledWith({ name: 'Test placemark 1' }, 0)
  })
})
