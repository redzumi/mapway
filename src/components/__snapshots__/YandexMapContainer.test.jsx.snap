// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YandexMapContainer renders correctly 1`] = `
.c0 {
  width: 600px;
  height: 400px;
  padding: 10px;
}

<YandexMapContainer
  mapCenter={
    Array [
      55.75120138894101,
      37.62371955158138,
    ]
  }
  onMapCenterUpdate={[MockFunction]}
  onPlacemarkCoordinatesUpdate={[MockFunction]}
  placemarks={Array []}
>
  <styled.div
    innerRef={[Function]}
  >
    <div
      className="c0"
    />
  </styled.div>
</YandexMapContainer>
`;
